rfcs/lower-bounds-calculation.md

```diff
@@ -0,0 +1,179 @@
+
+```lua
+function map_array(arr, f)
+    local result = {}
+    for i, e in ipairs(arr) do
+        table.insert(result, f(e, i, arr))
+    end
+    return result
+end
+
+local example = {1, 2, 3, 4}
+local example_result = map_array(example, function(i) return i * 2 end)
+```
+
+This function mirrors the actual `Array.map` function in JavaScript.  It is very frequent for users of this function to provide a lambda that only accepts one argument.  It would be annoying for callers to be forced to provide a lambda that accepts two unused arguments.  This obviously becomes even worse if the function later changes to provide yet more optional information to the callback.
+
+This use case is very important for Roblox, as we have many APIs that accept callbacks.  Implementors of those callbacks frequently omit arguments that they don't care about.
+
+Here is an example straight out of the Roblox developer documentation. ([full example here](https://developer.roblox.com/en-us/api-reference/event/BasePart/Touched))
+
+```lua
+local part = script.Parent
+
+local function blink()
+    -- ...
+end
+
+part.Touched:Connect(blink)
+```
+
+The `Touched` event actually passes a single argument: the part that touched the `Instance` in question.  In this example, it is omitted from the callback handler.
+
+We therefore want _oversaturation_ of a function to be allowed, but this combines with optional function arguments to create a problem with soundness.  Consider the following:
+
+```lua
+type Callback = (Instance) -> ()
+
+local cb: Callback
+function register_callback(c: Callback)
+    cb = c
+end
+
+function invoke_callback(i: Instance)
+    cb(i, e)
+end
+
+---
+
+function bad_callback(x: number?)
+end
+
+local obscured: () -> () = bad_callback
+
+register_callback(obscured)
+
+function good_callback()
+end
+
+register_callback(good_callback)
+```
+
+The problem we run into is, if we allow the subtyping rule `(T?) -> () <: () -> ()` and also allow oversaturation of a function, it becomes easy to obscure an argument type and pass the wrong type of value to it.
+
+We need a solution here.
+
+To resolve this, let's first examine the type of `bad_callback`.  It is tempting to type `bad_callback : (number?) -> number` and call it a day, but, if we read the definition of the Lua language, we're being somewhat hasty when we do that.  We can pass _as many extra arguments as we want, of any type_ to this function.  All save the first are ignored.  We could therefore instead decide to type `bad_callback : (number?, any...) -> number`.  We could also type `good_callback : (any...) -> ()`.
+
+So, we propose some different inference rules for functions:
+
+1. The AST fragment `function(arg0..argN) ... end` is typed `(T0..TN, any...) -> R` where `arg0..argN : T0..TN` and `R` is the inferred return type of the function body.  Function statements are inferred the same way.
```

Comment:
Is this specific to anonymous functions?

Reply:
No.  Named functions function exactly the same way.

Reply:
This feels somewhat problematic:

1. This means that even for direct calls, we're going to silently permit adding extra arguments everywhere; which feels like a notable reduction in strictness even if it's not a runtime error.
2. This means that all autocomplete / type hover / etc. is going to add the extra variadic annotation, which seems noisy and misleading since the function wasn't declared as variadic.
3. My understanding is that the only way to remove this is to ascribe the type to the *function*. This is arguably non-idiomatic in Luau, where we'd expect that to specify the function type, you use argument/return type annotations, but since there's no variadic argument you can't annotate it as not existing if that makes sense.

Let me know if I'm misunderstanding the proposal.

Reply:
1. Yes.  I mention that in the drawbacks section.  I think we solve it with a separate typechecking rule for function calls to functions whose exact definition is known.
2. Correct.  We could decorate function types with an extra bool to indicate whether or not this extra `any...` is implicit or explicit, perhaps?
3. We need to make this unobtrusive enough that people don't need to do this.

Reply:
I don't love this direction. The fact that we want to first relax the types and then add a separate rule makes me think we should not add `any...` by default, and instead specify the type checking rules in this RFC so that they don't result in the issues I mention.

I need to reread the RFC, but I'm wondering if we can separate a few concerns to make things line up:

- When inferring the constraints on function variables from calls, we infer the "lower bound" as described in this RFC
- When ascribing the type to a function definition, we use the semantics we have today (no implicit arguments that aren't present in source)
- When type checking calls where the function type is known, we require to match argument count in strict mode; potentially permitting any number in non-strict
- When casting function types to other types, we treat functions with fewer arguments as a supertype of functions with more arguments

Maybe I've just described the existing rules, unsure.

One other observation: the RFC claims that due to the call rules, it's more correct to ascribe a type with an extra variadic pack to function definitions.
I would claim that this is incorrect: from the runtime perspective, the *type* of the function that has 2 declared arguments is just that, a function with two arguments. It's observable through `debug.info` APIs if nothing else. What is *special* is not the type, it's the calling process, as such I think it's better to have the type of the definition reflect the declared arguments, but have the call/cast semantics reflect the balance we want between strictness/soundness/usability, which I tried to outline with strawman above.

Reply:
There is a clash here between examples like callback registration (where we want to allow the caller to supply more arguments than necessary) and examples like calling known functions (where we probably don't). E.g.

* `part.Touched:Connect(function(other) ... end)` OK
* `Vector3.new(1,2,3,4)` not OK?

Perhaps we need precise / imprecise function types in a similar fashion to precise / imprecise tables?

Reply:
@asajeffrey I wouldn't say that in a callback situation we "want to allow the caller to supply more arguments than necessary".
Caller of the callback knows precisely the type of function they are calling from their point of view:
```lua
function Instance:Connect(callback: (Instance, Instance) -> ())
    callback(one, other) -- no mismatch here and you can't provide extra arguments
end
```
The problem is with function type conversion/unification test happening during registration where `(Instance) -> ()` is passed where `(Instance, Instance) -> ()` is expected.

As a side note, existing bidirectional type inference can solve the case of `part.Touched:Connect(function(other) ... end)` by not only making `other` an `Instance` like it does right now, but by introducing other parameters while leaving them unnamed as well.

But unfortunately it cannot solve a case like this:
```lua
local function cb(other) ... end
part.Touched:Connect(cb)
```

Reply:
@asajeffrey Exactly right.  This is handled by the logic that typechecks function calls.  It is not quite the same as a subtype check.